subsys/net/ip/utils.c

```diff
@@ -486,32 +486,48 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
 #include <syscalls/net_addr_pton_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
-static uint16_t calc_chksum(uint16_t sum, const uint8_t *data, size_t len)
+/* Word based checksum calculation based on:
+ * https://blogs.igalia.com/dpino/2018/06/14/fast-checksum-computation/
+ * It’s not necessary to add octets as 16-bit words. Due to the associative property of addition,
+ * it is possible to do parallel addition using larger word sizes such as 32-bit or 64-bit words.
+ * In those cases the variable that stores the accumulative sum has to be bigger too.
+ * Once the sum is computed a final step folds the sum to a 16-bit word (adding carry if any).
+ */
+uint16_t calc_chksum(uint16_t sum_in, const uint8_t *data, size_t len)
 {
-	const uint8_t *end;
-	uint16_t tmp;
-
-	end = data + len - 1;
+	uint64_t sum = (uint64_t)ntohs(sum_in);
+	uint32_t *p = (uint32_t *) data;
```

Comment:
What about alignment? This is not gonna work on all platforms, is it?

Reply:
The sum is maintained in host endiannes, while the data is always big endian notation. By converting to big endian at the start, the whole checksum can be computed in big endian and at the end you go back to host endianness. So I think it should work both on big en little endian systems, I only tested on little endian so far.

Reply:
Could you have a look if you are satisfied with the latest changes?

Reply:
@carlescufi: Are you ok with the latest implementation?